contracts/exchangeIssuance/ExchangeIssuance.sol

```diff
@@ -0,0 +1,768 @@
+     * 
+     * @param _setToken     The SetToken that is being liquidated
+     * @return              Amount of WETH received after liquidating all components of the SetToken
+     */
+    function _liquidateComponentsForWETH(ISetToken _setToken) internal returns (uint256) {
+        uint256 sumEth = 0;
+        ISetToken.Position[] memory positions = _setToken.getPositions();
+        for (uint256 i = 0; i < positions.length; i++) {
+            address token = positions[i].component;
+            uint256 tokenBalance = IERC20(token).balanceOf(address(this));
+            
+            // Get max amount of WETH for the available amount of SetToken component
+            (, Exchange exchange) = _getMaxTokenForExactToken(tokenBalance, token, WETH);
+            sumEth = sumEth.add(_swapExactTokensForTokens(exchange, token, WETH, tokenBalance));
+        }
+        return sumEth;
+    }
+    
+    /**
+     * Issues SetTokens for an exact amount of input WETH. 
+     * Acquires SetToken components at the best price accross uniswap and sushiswap.
+     * Uses the acquired components to issue the SetTokens.
+     * 
+     * @param _setToken         Address of the SetToken being issued
+     * @param _minSetReceive    Minimum amount of index to receive
+     * @return setTokenAmount   Amount of SetTokens issued
+     */
+    function _issueSetForExactWETH(ISetToken _setToken, uint256 _minSetReceive) internal returns (uint256) {
+        uint256 wethBalance = IERC20(WETH).balanceOf(address(this));
+        
+        ISetToken.Position[] memory positions = _setToken.getPositions();
+        
+        (uint256[] memory amountEthIn, Exchange[] memory exchanges, uint256 sumEth) = _getAmountETHForIssuance(_setToken);
+
+        uint256 setTokenAmount = _acquireComponents(positions, amountEthIn, exchanges, wethBalance, sumEth);
+        
+        require(setTokenAmount > _minSetReceive, "ExchangeIssuance: INSUFFICIENT_OUTPUT_AMOUNT");
+        
+        basicIssuanceModule.issue(_setToken, setTokenAmount, msg.sender);
+        
+        return setTokenAmount;
+    }
+    
+    /**
+     * Issues an exact amount of SetTokens using WETH. 
+     * Acquires SetToken components at the best price accross uniswap and sushiswap.
+     * Uses the acquired components to issue the SetTokens.
+     * 
+     * @param _setToken          Address of the SetToken being issued
+     * @param _amountSetToken    Amount of SetTokens to issue
+     * @return sumEth            Total amount of ether used to acquire the SetToken components    
+     */
+    function _issueExactSetFromWETH(ISetToken _setToken, uint256 _amountSetToken) internal returns(uint256) {
+        
+        uint256 sumEth = 0;
+        
+        ISetToken.Position[] memory positions = _setToken.getPositions();
+        for(uint256 i = 0; i < positions.length; i++) {
+            
+            uint256 amountToken = uint256(positions[i].unit).mul(_amountSetToken).div(1 ether);
```

Comment:
Please use `preciseMul` - which looks like `uint256(positions[i].unit).preciseMul(_amountSetToken)`

You'll need to pull in this file and do `using PreciseUnitMath for uint256`
https://github.com/SetProtocol/index-coop-contracts/blob/master/contracts/lib/PreciseUnitMath.sol

Reply:
Also please update this for all other parts of the codebase